Auto-Sync-Sail blogpost (#68)
rizinorg/website: content/posts/gsoc-2024-auto-sync-sail/index.md

Hello, I’m Mostafa. I graduated with Excellence from Cairo University’s Faculty of Engineering, Computer Engineering Department, class of 2023. I write C++ for a living. I love systems programming, metaprogramming & DSLs, as well as Compilers & VMs. You can find me @[Github](https://github.com/moste00), and @[Linkedin](www.linkedin.com/in/mostafa-m-kamal-b015691b1).

> I was honored to join GSoC in 2024, for my second (and last!) time, @Rizin org. The project that I applied for was the binary lifting of RISC-V instructions to Rizin’s bespoke internal representation, RzIL. However, updating the RISC-V capstone disassembler (originally a small task in the project preamble) turned out to need much more work than expected, and blocked the rest of the project.

Reviewer: ```suggestion
I was honored to participate again as a student in the 2024 GSoC with the Rizin Organization. The original project was implementing binary lifting techniques for RISC-V instructions onto their custom internal representation called RzIL. However, updating the RISC-V capstone disassembler (originally a small task in the project preamble) turned out to need much more work than expected, and blocked the rest of the project.
```

Let's start at the beginning.

# RISC-V… Lifting?
***Lifting*** is a term of art in compiler research and implementation, it refers to any process that takes as input a low-level machine code program and outputs a higher-level program. The reverse process, ***Lowering***, is what compilers do when they compile from a relatively high level language like C or LLVM IR to machine code. So you could simply think of Lifting as a synonym for “Decompiling” or “Reverse-Compiling”. 

In the context of Rizin, Lifting refers to transforming a machine code program written for any of the hardware architectures that Rizin understands (x86, RISC-V, 6502, etc...) to a Rizin-specific intermediate language called RzIL. 

> By doing this, Rizin’s developers can write generic analysis algorithms that manipulate RzIL instructions, and a generic VM that executes Rizin instructions, only once. Then, for each architecture that Rizin supports, a lifter that transforms machine code written for that architecture into RzIL is written, and we get all the analysis algorithms and VM execution capabilities “for free”. 

Reviewer 2: ```suggestion
By doing this, Rizin’s developers can write generic analysis algorithms that interpret RzIL instructions, and a generic VM that executes them, only once. Then, for each architecture that Rizin supports, a lifter that transforms machine code written for that architecture into RzIL is written, and we get all the analysis algorithms and VM execution capabilities “for free”. 
```

Maybe a bit shorter and less repetition of "instruction"?

In a nutshell, RzIL is the universal “Lingua Franca” for Rizin, like English is for Software Engineering.

![IMAGE__WORLD_WITHOUT_RZIL](https://hackmd.io/_uploads/HyfLVzMHkg.png)
Figure 1: Without RzIL, there is no smarter way to perform N operations for M assembly languages other than doing an NxM amount of work, implementing the N operations over and over again per each language/architecture.

> ![IMAGE__WORLD_WITH_RZIL](https://hackmd.io/_uploads/HklxSGfHyg.png)

Reviewer: please push all the pngs in the same folder as the index.md

then fix the links like in the example below: https://github.com/rizinorg/website/blob/fc869385e850146f02cc36697b38f27402f2b6ce/content/posts/gsoc-2021-summary/index.md?plain=1#L120

Figure 2: With RzIL, the amount of work to support N operations for M architectures is N+M, the N operations are written exactly once for the intermediate language, then M transformers are written to lift each of the M architectures to the intermediate language.

# For want of a disassembler

So the original plan was to write the grey arrow in the figure above: a lifter from RISC-V machine code into RzIL. However, the first step in doing that is to “parse” RISC-V instructions from their binary form into a convenient data structure. We call that “parsing” step disassembly, or, more accurately, decoding.

>> Side Note:  lots of people, when “disassembly” and “assembly” are mentioned, will probably think of the following diagram:
![IMAGE__ASM_DISASM](https://hackmd.io/_uploads/Sy9mIfzHkl.png)
This is not wrong for most purposes. However, in the context of this writeup it’s better to have the following and more detailed picture in mind:
![IMAGE__ASM_DISASM_ENC_DEC](https://hackmd.io/_uploads/BySYvMfByx.png)

> In this writeup I’m more interested in the left-to-right flow: decoding from a binary to a structured (e.g. C struct) representation of the instruction, then assembling the structured representation of the instruction into a string form. Confusingly, sometimes “Disassembly” is used to include both Disassembly **and** Decoding, for example in Capstone the structured representation includes as a member its own toString serialization. It will often be clear from context what step is meant, and decoding is often far more important than disassembly. 

Reviewer 2: ```suggestion
In this writeup I’m more interested in the left-to-right flow: decoding from a binary to a structured (e.g. C struct) representation of the instruction, then assembling the structured representation of the instruction into a string form. Confusingly, sometimes “Disassembly” is used to include both Disassembly **and** Decoding, for example in Capstone the structured representation includes as a member its own `toString` serialization. It will often be clear from context what step is meant, and decoding is often far more important than disassembly. 
```

Where were we? Ah yes, we were supposed to “parse” (i.e. decode) an instruction from its binary form into a convenient data structure, so that we can write elegant code that easily and robustly lifts it into RzIL. 

The good news is that Rizin already has a RISC-V decoder/disassembler, since it uses as a library the project *Capstone*, which is a general-purpose disassembler framework for multiple architectures, including RISC-V.

> The bad news? Those disassemblers are often incomplete and out of date. 

Reviewer 2: ```suggestion
The bad news? The RISC-V disassembler was incomplete and out of date. 
```

Please :)
We worked hard for the others :)

Capstone was originally written based on codegen logic from LLVM. It’s essentially a port of LLVM disassembly logic from C++ to C (along with much simplification and cleaning up). Unfortunately, LLVM keeps updating that logic to reflect the fast-moving development and evolution of the architectures; those updates are not magically reflected back into Capstone! To make matters even worse, even LLVM proper can’t completely keep up with all the updates that happen to all the architectures it supports, it lags.

The Capstone project maintains an [update tool](https://rizin.re/posts/auto-sync/) called [Auto-Sync](https://github.com/capstone-engine/capstone/blob/next/suite/auto-sync/intro.md), which can semi-automatically synchronize changes from LLVM to Capstone (using Tree-sitter magic). Alas, it can only do that for some architectures, and RISC-V is not among the supported ones. Also, we already saw how even LLVM is not completely on top of all updates. Fortunately, the solution exists, just hiding elswhere.

> # To Sail the high seas and RISC it all

Reviewer 2: xD

Let’s see a snippet of what Sail looks like in practice, here’s the definition of RISC-V IType (immediate) instructions:
![IMAGE__SAIL_SNIPPET_ITYPE_ENCDEC](https://hackmd.io/_uploads/ryhunMMBkg.png)

The rule might be as cryptic as latin if you’re not used to pattern-matching constructs from functional languages, but what it’s saying is simply the following:
1. If the first (least-significant) 7 bits of the 32-bit instruction are: 0010011
1. And if the 3 bits from bit 12 to bit 14 are in the table encdec_iop
1. Then, the operation specified by this instruction is whatever enum corresponding to bits 12:14 in the encdec_iop table, and the args are:
    * The 5-bit register index **rd** in bits 7 through 11
    * The 5-bit register index **rs1** in bits 15 through 19
    * The 12-bit literal **imm** in bits 20 through 31

> 4. Otherwise, if (1) and (2) are not true, keep checking the 32-bit binary instruction against other rules

Reviewer 2: Which line does this in he source above?

Author: If you mean the destructuring of the instruction into [`rd, rs1, imm`], it's implicit in the way the mapping pattern is constructed (last line in the image), `@` is the vector concatenation operator in Sail. The actual bit widths are inferred because the types of `rd, rs1,` and `imm` were all declared before in the AST clause declaration (first line, `regidx` was declared earlier in the project as a synonym for a bitvector of width 5).

If you mean point (4), that's just pattern matching proceeding sequentially across the clauses of `encdec` until it succeeds. The only difference is that Sail has a unique and innovative feature of "Scattered Mappings", which means you can write different rules for `encdec` across different lines and different source files (separated by other logic and definitions) and Sail will still consider all of them part of the same logical pattern-match expression.

Reviewer 2: Yes, I meant point 4. But isn't `ecndec` not already described in point 2?
Because one can't see these other `encdec` definitions in the screenshot, I think it is important to mention/explain the scattered mapping or cut it for brevity.

The [generator](https://github.com/rizinorg/capstone-autosync-sail) itself is ~2500 lines of idiomatic (I hope) OCaml. But since Sail is a complex language, the tool must make some assumptions about the input that might not survive the evolution of Sail-RISCV, the application to other Sail models, or the active evolution of Sail itself. 

Finally, we haven’t addressed the original problem yet! I hope to eventually and finally write the grey arrow in the first diagram: Lifting RISC-V instructions to RzIL code.

# Conclusion

Let’s summarize this rollercoaster journey:

1. We just wanted to write a binary lifter for RISC-V instructions into RzIL, Rizin’s intermediate language
1. But in order to do that, we first have to have an up-to-date RISC-V decoder/disassembler
1. Rizin depends on Capstone for RISC-V disassembly, but Capstone RISC-V disassembly logic is ported from old LLVM logic that is not up-to-date
1. Even modern LLVM is not completely up-to-date with RISC-V
1. But Sail-RISCV is, and it's adopted by the RISC-V foundation as the most authoritative model of the RISC-V architecture

> 1. And thus, we can generate a Capstone disassembler module from Sail-RISCV, by depending on the Sail compiler as a library

Reviewer 2: Please add '.' at the end of each sentence.

It was fun. Frustrating and long-winded at times, but what kind of programming isn’t? That’s part of the thrill anyway!

> That's all and Happy Christmas! Keep coding through the wind and the snow. 

Reviewer: ```suggestion
That's all and Happy Holidays! Keep coding through the wind and the snow. 
```
